### PR TITLE
Introduce predicate parameter for conditional exception mapping

### LINKED ISSUE #130: Map Exception condition
Dear Kristian,

If we want to Map for a specific exception state (like `catch (SomeException ex) when (ex.Prop == "something")`) we cannot do that now. I only find that we can return `null`, or call `Rethrow` for an exception but this misses out the other Map configurations.

I think it will be a good enhancement if we can give a predicate where we define the condition of the Map but not the middleware.

Currently:
```cs
c.Map<DbUpdateException>((ctx, ex) =>
{
    var problemDetailsFactory = ctx.RequestServices
        .GetRequiredService<MvcProblemDetailsFactory>();

    if (ex.InnerException is SqlException sqlException)
    {
        // Unique constraints
        if (sqlException.Number == 2601 || sqlException.Number == 2627)
        {
            return problemDetailsFactory
                .CreateProblemDetails(ctx, StatusCodes.Status409Conflict, "Constraint Error");
        }
    }

    // this code duplication sould not be here
    return problemDetailsFactory
        .CreateProblemDetails(ctx, StatusCodes.Status500InternalServerError, "Other Error");
});

// everything else
c.Map<Exception>((ctx, ex) =>
    ctx.RequestServices
        .GetRequiredService<MvcProblemDetailsFactory>()
        .CreateProblemDetails(ctx, StatusCodes.Status500InternalServerError, "Other Error"));
```

My API suggestion is the following

```cs
c.Map<DbUpdateException>(
    ex => ex.InnerException is SqlException sqlException && (sqlException.Number == 2601 || sqlException.Number == 2627),
    (ctx, ex) =>
    {
        var problemDetailsFactory = ctx.RequestServices
            .GetRequiredService<MvcProblemDetailsFactory>()
            .CreateProblemDetails(ctx, StatusCodes.Status409Conflict, "Constraint Error");
    }
);

// everything else
c.Map<Exception>((ctx, ex) =>
    ctx.RequestServices
        .GetRequiredService<MvcProblemDetailsFactory>()
        .CreateProblemDetails(ctx, StatusCodes.Status500InternalServerError, "Other Error"));
```

Regards

### PR DESCRIPTION
**Pull request by tibitoth**
resolves #130 

@khellang I've implemented the API suggestion from the linked issue. The new test probably not the best looking but it covers the logic.

### CODE REVIEW

File: src/ProblemDetails/ProblemDetailsOptions.cs [hunk 1]
```diff
@@ -177,13 +177,38 @@ public ProblemDetailsOptions()
         /// <remarks>
         /// Mappers are called in the order they're registered.
         /// Returning <c>null</c> from the mapper will signify that you can't or don't want to map the exception to <see cref="ProblemDetails"/>.
-        /// This will cause cause the exception to be rethrown.
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsOptions.cs:180]: Nice 😍

File: src/ProblemDetails/ProblemDetailsOptions.cs [hunk 2]
```diff
@@ -288,24 +313,27 @@ internal bool TryMapProblemDetails(HttpContext context, Exception? exception, ou
...
+
             public bool CanMap(Type type)
             {
                 return Type.IsAssignableFrom(type);
             }
 
             public bool TryMap(HttpContext context, Exception exception, out MvcProblemDetails? problem)
             {
-                if (CanMap(exception.GetType()))
+                if (CanMap(exception.GetType()) && Predicate(context, exception))
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsOptions.cs:336]: It feels cleaner to move the `Predicate` invocation into `CanMap`, maybe rename it to `ShouldMap`? 😅

File: src/ProblemDetails/ProblemDetailsOptions.cs [hunk 3]
```diff
@@ -288,24 +313,27 @@ internal bool TryMapProblemDetails(HttpContext context, Exception? exception, ou
 
         private sealed class ExceptionMapper
         {
-            public ExceptionMapper(Type type, Func<HttpContext, Exception, MvcProblemDetails?> mapping)
+            public ExceptionMapper(Type type, Func<HttpContext, Exception, MvcProblemDetails?> mapping, Func<HttpContext, Exception, bool> predicate)
             {
                 Type = type;
                 Mapping = mapping;
+                Predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsOptions.cs:320]: Can we throw for the other parameters as well? We obviously don't support passing `null` 😄

**Review by khellang — CHANGES_REQUESTED** [review]
This looks really good, just a few nits. Thanks again! 🙏🏻

File: src/ProblemDetails/ProblemDetailsOptions.cs (hunk 2 above)
Comment (tibitoth) [reply src/ProblemDetails/ProblemDetailsOptions.cs]: done

File: src/ProblemDetails/ProblemDetailsOptions.cs (hunk 3 above)
Comment (tibitoth) [reply src/ProblemDetails/ProblemDetailsOptions.cs:320]: done

**Comment by tibitoth** [comment]
> This looks really good, just a few nits. Thanks again! 🙏🏻

Thanks for the review ;)

**Review by khellang — APPROVED** [review]

**Comment by khellang** [comment]
Love it! 🥰 I'll try to get out a new version on NuGet ASAP 😊

**Comment by khellang** [comment]
New version is on NuGet 😊

**Pull request merged**